docs/content/en/docs/concepts/api/_index.md

```diff
@@ -0,0 +1,71 @@
+---
+title: "Skaffold API"
+linkTitle: "Skaffold API"
+weight: 40
+---
+
+When using `skaffold dev` or `skaffold debug`, Skaffold exposes a API server over its lifetime.
+The API server is the primary way tools like IDEs to integrate with skaffold and subscribe to events 
+regarding the different phases in a pipeline run and to get the overall state of the pipeline.
+The API server also provides fine grain controls to Skaffold's individual components:
+build, deploy and sync, as opposed to relying on Skaffold’s built-in trigger mechanisms.
+
+
+## Skaffold API 
+Skaffold API is restful and `gRPC` based, so it works with any language that has an HTTP library, such as cURL and urllib.
+The API server runs on localhost at predefined ports.
+The protos used can be found here. (todo add link)
+
+### gRPC Server
+
+gRPC API server is exposed on port `50051` by default. If the port is busy, Skaffold will find the next available port. 
+You can grab the port from Skaffold logs.
+
+```code
+$ skaffold dev
+WARN[0000] port 50051 for gRPC server already in use: using 50053 instead 
+``` 
+You can also specify a port on the command line with flag `--rpc-port`.
+
+
+### gRPC REST Server  
+REST API server is exposed on port `50052` by default. If the port is busy, Skaffold will find the next available port. 
+You can grab the port from Skaffold logs.
+
+```code
+$ skaffold dev
+WARN[0000] port 50052 for gRPC HTTP server already in use: using 50055 instead 
+``` 
+You can also specify a port on the command line with flag `--rpc-http-port`.
+
+
+## Skaffold API
+Skaffold API Server exposes following endpoints.
+
+### GET /v1/events
+
+Skaffold provides a continuous development mode [`skaffold dev`](../modes/#skaffold_dev) which builds, deploys
```

Comment:
It is not clear enough what we get as response of this call. I suppose it is a "long polling" response similar to the "watch" operation on Kubernetes, for example. Is it the case? If so, is it using chunk transfer encoding, or something else?

Reply:
We will link the protos soon and hope it will make it more clear.

Reply:
we could add more info about this in here: 

on the gRPC layer you have streaming gRPC of the event stream (protos will explain). 
for the HTTP layer we use https://github.com/grpc-ecosystem/grpc-gateway they translate gRPC streams to newline separated JSON stream which is using chunk transfer encoding. See an example curl output with `--rpc-http-port 5555`: 

curl -v http://localhost:5555/v1/events
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 5555 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 5555 (#0)
> GET /v1/events HTTP/1.1
> Host: localhost:5555
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Wed, 16 Oct 2019 20:45:37 GMT
< Transfer-Encoding: chunked
< 
{"result":{"timestamp":"2019-10-16T20:45:15.402079Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version:v0.39.0-131-g1759410a7-dirty ConfigVersion:skaffold/v1beta16 GitVersion: GitCommit:1759410a7071b1b7858a2a8b85f8a8bfb07212da GitTreeState:dirty BuildDate:2019-10-10T11:44:31Z GoVersion:go1.13.1 Compiler:gc Platform:darwin/amd64}"}}}}
{"result":{"timestamp":"2019-10-16T20:45:15.847029Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.847022Z","event":{"buildEvent":{"artifact":"node-guestbook-frontend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-frontend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.859036Z","event":{"buildEvent":{"artifact":"node-guestbook-frontend","status":"Complete"}},"entry":"Build completed for artifact node-guestbook-frontend"}}
{"result":{"timestamp":"2019-10-16T20:45:15.861777Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"In Progress"}},"entry":"Build started for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:16.316248Z","event":{"buildEvent":{"artifact":"node-guestbook-backend","status":"Complete"}},"entry":"Build completed for artifact node-guestbook-backend"}}
{"result":{"timestamp":"2019-10-16T20:45:16.326313Z","event":{"deployEvent":{"status":"In Progress"}},"entry":"Deploy started"}}
{"result":{"timestamp":"2019-10-16T20:45:17.605745Z","event":{"deployEvent":{"status":"Complete"}},"entry":"Deploy complete"}}
```

Reply:
We should try to be clear about a few things for all of the endpoints here:

* the exact URL, and the HTTP method
* the expected response (both proto and JSON)
* the expected payload, if any (proto and JSON examples would be helpful here too)

Also, for endpoints like this one, implementation details are useful. For example, when using the /v1/events endpoint over HTTP, it's useful to know that it's using opening a long standing connection (aka the connection is open until the client cancels or the server shuts down), and it's using HTTP/2 server push to send events to an open connection ala pub sub.

Reply:
Regarding response for, i will link in the EventLog proto from the api ref which just got merged.
i think the payload will come we write the docs for `/v1/execute`

Reply:
in an API spec we need to be precise with the response. 
The EventLog proto will only refer to the gRPC doc, not the JSON format, and definitely doesn't have information about how streaming is implemented for HTTP clients. We need to have that here as well.